### PR TITLE
Fix for AEP Ohio data download home-assistant/core#112744

### PR DESCRIPTION
**Pull request by drewclauson**
Fix for AEP Ohio data download home-assistant/core#112744

### CODE REVIEW

**Comment by drewclauson** [comment]
@tronikos @benhoff I'm open to all feedback, since it's been a long time since I've done anything Python.  Thanks!

**Comment by benhoff** [comment]
Doesn't `@abstractmethod` force you to implement the method in every subclass, or it will break the subclass and prevent it from being instantiated in the first place, right?

This commit would break the other implementations/subclasses of AEPBase if that is correct.

I'm a bit fuzzy on Abstract Base Class and some of these speciality decorators.

**Comment by benhoff** [comment]
I'm also for removing the extra if statement under async_login, but you'll see the assertion error from [this issue](https://github.com/home-assistant/core/issues/110019) unless every subclass of AEPBase has the cookie to be removed set.

I'll try to run this tomorrow and confirm it works, code overall looks good to me other than that.

**Comment by drewclauson** [comment]
> Doesn't `@abstractmethod` force you to implement the method in every subclass, or it will break the subclass and prevent it from being instantiated in the first place, right?
> 
> This commit would break the other implementations/subclasses of AEPBase if that is correct.
> 
> I'm a bit fuzzy on Abstract Base Class and some of these speciality decorators.

> I'm also for removing the extra if statement under async_login, but you'll see the assertion error from [this issue of "async_login not called"](https://github.com/home-assistant/core/issues/110019) unless every subclass of AEPBase has the "cookie to be removed" set or implemented.
> 
> Having had to backtrack the above issue before, it's not a very helpful or accurate error code.
> 
> I'll try to run this branch tomorrow and confirm it works, code overall looks good to me other than that.


I'm more fuzzy than you I think, I was mostly looking for something on how to make the method optional... and pass the pre-commit run!  So I read the docs wrong.  I'm working on a change to that method to keep other implementations from failing without implementing that method.  The problem I ran into was the linter did not like working with Optional[list[str]] (because I wanted to be all fancy and allow multiple domains to be cleared, but I'll experiment with it more!

I've been so buried in Java/TypeScript lately that the Python syntax right now just looks like Latin to me, so thanks for the feedback, it's helping me learn. :-)

**Comment by drewclauson** [comment]
I made some tweaks.  Theoretically... if we are clearing cookies, then we shouldn't need the extra if statement that was added under the issue of "async_login not called", so I added a check for None on `cls.login_cookies_to_clear()` before that block and added it back in so it'll continue to work for the other implementations.  Take a look!

File: src/opower/utilities/aepbase.py [hunk 1]
```diff
@@ -112,6 +117,12 @@ async def async_login(
         optional_mfa_secret: Optional[str],
     ) -> None:
         """Login in AEP using user/pass and then do the SAML call to opower."""
+        # Clear cookies before logging in again, in case old ones are still around
+        if cls.login_cookies_to_clear() is not None:
```
Comment (tronikos) [inline src/opower/utilities/aepbase.py:121]: Revert all other changes (except removal of `if not login_parser.password_field_found:` block) and just call here:
`session.cookie_jar.clear(lambda cookie: cookie["domain"].endswith(".opower.com"))`

Comment (drewclauson) [reply src/opower/utilities/aepbase.py]: Done, makes it simple, thanks

**Comment by drewclauson** [comment]
Updated PR at maintainer's request.  Makes it much simpler.

**Comment by benhoff** [comment]
I tested this, and it works for me. Code looks good.

**Review by tronikos — APPROVED** [review]

**Comment by tronikos** [comment]
Released. Please send a PR to bump the version on the Home Assistant side.

**Pull request merged**